client/qnr.go

```diff
@@ -0,0 +1,130 @@
+package client
+
+import (
+	"errors"
+	"fmt"
+	"github.com/xlab-si/emmy/common"
+	"github.com/xlab-si/emmy/dlog"
+	pb "github.com/xlab-si/emmy/protobuf"
+	"github.com/xlab-si/emmy/qrproofs"
+	"google.golang.org/grpc"
+	"math/big"
+)
+
+type QNRClient struct {
+	genericClient
+	prover  *qrproofs.QNRProver
+	variant pb.SchemaVariant
+}
+
+func NewQNRClient(conn *grpc.ClientConn, qr *dlog.QR, y *big.Int) (*QNRClient, error) {
+	genericClient, err := newGenericClient(conn)
+	if err != nil {
+		return nil, err
+	}
+
+	return &QNRClient{
+		genericClient: *genericClient,
+		prover:        qrproofs.NewQNRProver(qr, y),
+	}, nil
+}
+
+// Run starts protocol for proving that y is QNR.
+func (c *QNRClient) Run() (bool, error) {
+	c.openStream()
+	defer c.closeStream()
+	m := c.prover.QR.N.BitLen()
+
+	// message where y is sent to the verifier
+	initMsg := &pb.Message{
+		ClientId:      c.id,
+		Schema:        pb.SchemaType_QNR,
+		SchemaVariant: pb.SchemaVariant_SIGMA,
+		Content: &pb.Message_Bigint{
+			&pb.BigInt{X1: c.prover.Y.Bytes()},
+		},
+	}
+
+	resp, err := c.getResponseTo(initMsg) // simply an empty message
+	if err != nil {
+		return false, err
+	}
+	if resp.GetEmpty() == nil {
+		return false, fmt.Errorf("should receive empty message")
+	}
+
+	proved := false
+	// the client has to prove for all i - if in one iteration the knowledge
+	// is not proved, the protocol is stopped
+	for i := 0; i < m; i++ {
+		msg := &pb.Message{
```

Comment:
Changed. Note that in Schnorr the naming is wrong - getProofRandomData actually sends proofRandomData and gets challenge (PedersenCommitment).